### PR TITLE
[inductor] support vec for atomic add

### PR DESCRIPTION
**Pull request by zhuhaozhe**
Depends on https://github.com/pytorch/pytorch/pull/130827 to have correct `index_expr` dtype

Support vec for atomic add by scalar implementation.
TestPlan:
```
python test/inductor/test_cpu_repro.py -k test_scatter_using_atomic_add_vec
```
Generated code for `test_scatter_using_atomic_add_vec`
```
cpp_fused_scatter_0 = async_compile.cpp_pybinding(['const float*', 'const int64_t*', 'const float*', 'float*'], '''
#include "/tmp/torchinductor_root/nn/cnnpkaxivwaa5rzng6qsyc4ao42vschogi3yk33ukwv3emlvxeqq.h"
extern "C"  void kernel(const float* in_ptr0,
                       const int64_t* in_ptr1,
                       const float* in_ptr2,
                       float* out_ptr0)
{
    {
        for(long x0=static_cast<long>(0L); x0<static_cast<long>(16L); x0+=static_cast<long>(16L))
        {
            auto tmp0 = at::vec::Vectorized<float>::loadu(in_ptr0 + static_cast<long>(x0), 16);
            tmp0.store(out_ptr0 + static_cast<long>(x0));
        }
        #pragma omp simd simdlen(8) 
        for(long x0=static_cast<long>(16L); x0<static_cast<long>(25L); x0+=static_cast<long>(1L))
        {
            auto tmp0 = in_ptr0[static_cast<long>(x0)];
            out_ptr0[static_cast<long>(x0)] = tmp0;
        }
    }
    {
        for(long x0=static_cast<long>(0L); x0<static_cast<long>(16L); x0+=static_cast<long>(16L))
        {
            auto tmp0 = at::vec::VectorizedN<int64_t,2>::loadu(in_ptr1 + static_cast<long>(x0), 16);
            auto tmp12 = at::vec::Vectorized<float>::loadu(in_ptr2 + static_cast<long>(x0), 16);
            auto tmp1 = 25L;
            auto tmp2 = c10::convert<int64_t>(tmp1);
            auto tmp3 = at::vec::VectorizedN<int64_t,2>(tmp2);
            auto tmp4 = tmp0 + tmp3;
            auto tmp5 = static_cast<int64_t>(0);
            auto tmp6 = at::vec::VectorizedN<int64_t,2>(tmp5);
            auto tmp7 = at::vec::VecMask<int64_t,2>(tmp0 < tmp6);
            auto tmp8 = decltype(tmp4)::blendv(tmp0, tmp4, tmp7.template cast<int64_t,2>());
            auto tmp9 =
            [&]
            {
                __at_align__ std::array<int64_t, 16> tmpbuf;
                tmp8.store(tmpbuf.data());
                return tmpbuf;
            }
            ()
            ;
            auto tmp10 =
            [&]
            {
                __at_align__ std::array<int64_t, 16> tmpbuf;
                #pragma GCC unroll 16
                for (long x0_inner = 0; x0_inner < 16; x0_inner++)
                {
                    tmpbuf[x0_inner] = static_cast<long>(tmp9[x0_inner]);
                }
                return at::vec::VectorizedN<int64_t,2>::loadu(tmpbuf.data(), 16);
            }
            ()
            ;
            TORCH_CHECK((at::vec::VecMask<int64_t,2>((at::vec::VectorizedN<int64_t,2>(0) <= tmp10) & (tmp10 < at::vec::VectorizedN<int64_t,2>(25L)))).all_masked(), "index out of bounds: 0 <= tmp10 < 25L");
            atomic_add_vec(out_ptr0, tmp8, tmp12);
        }
        #pragma omp simd simdlen(8) 
        for(long x0=static_cast<long>(16L); x0<static_cast<long>(20L); x0+=static_cast<long>(1L))
        {
            auto tmp0 = in_ptr1[static_cast<long>(x0)];
            auto tmp9 = in_ptr2[static_cast<long>(x0)];
            auto tmp1 = 25L;
            auto tmp2 = c10::convert<int64_t>(tmp1);
            auto tmp3 = decltype(tmp0)(tmp0 + tmp2);
            auto tmp4 = tmp0 < 0;
            auto tmp5 = tmp4 ? tmp3 : tmp0;
            auto tmp6 = tmp5;
            auto tmp7 = c10::convert<int64_t>(tmp6);
            TORCH_CHECK((0 <= tmp7) & (tmp7 < 25L), "index out of bounds: 0 <= tmp7 < 25L");
            atomic_add(&out_ptr0[static_cast<long>(tmp5)], static_cast<float>(tmp9));
        }
    }
}
''')
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by jgong5 — COMMENTED** [review]
If it depends on the change from another PR, is it possible to stack it on that change to simplify the code review? Thanks.

**Comment by jgong5** [comment]
Could you share the perf numbers with the change?

**Comment by zhuhaozhe** [comment]
> If it depends on the change from another PR, is it possible to stack it on that change to simplify the code review? Thanks.

That PR is merged, I just rebased this PR

**Comment by zhuhaozhe** [comment]
> Could you share the perf numbers with the change?

Overall speedup on all suites are similar.
3 models regressions and 5 models improved. One of the diffs on regressed kernel:
image

Hi, @jgong5. Do you have suggestions on it? 
The kernel is vectorized but 2x slower, should we consider heuristic to disable such case?

File: torch/_inductor/codegen/cpp.py
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2319]: I guess the intention is to vectorize the index? But it is still possible that the index is a scalar in which case we should have a specialized version of `atomic_add_vec` for it?

File: test/inductor/test_cpu_repro.py
Comment (jgong5) [inline test/inductor/test_cpu_repro.py:1966]: Since we have a special path for `num_threads == 1`, shall we have UTs covering both single-threaded and multi-threaded cases?

**Review by jgong5 — CHANGES_REQUESTED** [review]

**Comment by jgong5** [comment]
> Hi, @jgong5. Do you have suggestions on it?
> The kernel is vectorized but 2x slower, should we consider heuristic like #122961 to disable such case?

Do all the three regressed cases have similar code pattern? We can collect all the patterns and identify the solution.

**Comment by zhuhaozhe** [comment]
3 

> > Hi, @jgong5. Do you have suggestions on it?
> > The kernel is vectorized but 2x slower, should we consider heuristic like #122961 to disable such case?
> 
> Do all the three regressed cases have similar code pattern? We can collect all the patterns and identify the solution.

3 regressed model from TAS: `GPTJForCausalLM`, `basic_gnn_gin`, `basic_gnn_sage`.
`basic_gnn_gin`, `basic_gnn_sage` are same case.
For `GPTJForCausalLM` I do not reproduced the regression and the generated code is same, so I think it should be a false positive.

**Comment by zhuhaozhe** [comment]
From the vtune analyzer, the `loads blocked by store forwarding` get more serious after vectorized, since we only meet this regression for `low precision dtypes`, considered that the difference between `lowp` and `float32` is we are using half bits of a Vec and use `masked load/store` instead of `load/store` intrinsics this should be the reason.
So I tried to apply this PR https://github.com/pytorch/pytorch/pull/126502 to use full bits load/store with `lowp` and the previous regression case disappear
### Previous regression case:
image

After that, we found a new regression loop body: ( <5% end2end, ~25% for this kernel only)
### New found regression
```
![image](https://github.com/user-attachments/assets/4070f1a9-6ccf-4903-a962-a3c7e3b3a275)

```
This regression should be caused by introduce too many non-contiguous load/store after vec is supported for atomic_add and it should be addressed by this PR https://github.com/pytorch/pytorch/pull/122961 (disable vec if there are too many non-contiguous in a loop body)

File: torch/_inductor/codegen/cpp.py
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2115]:
```suggestion
        acc_store: bool = False,
```

**Review by jgong5 — CHANGES_REQUESTED** [review]
Please address my other comments.

File: torch/_inductor/codegen/cpp.py
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py]: Updated, thanks

File: test/inductor/test_cpu_repro.py
Comment (zhuhaozhe) [reply test/inductor/test_cpu_repro.py:1966]: Thanks for advices, added.

File: torch/_inductor/codegen/cpp.py
Comment (zhuhaozhe) [reply torch/_inductor/codegen/cpp.py:2284]: The intension here is to reuse `index_expr` to load contiguous or not contiguous index.
Although `index_expr` support to handle scalar index,  I don't think the index here should not be a scalar 
The original code use `_get_store_line` here and looks like `_get_store_line` assume index will not be a scalar:
```
    def _get_store_line(
        self,
        value: Union[str, CppCSEVariable],
        var: str,
        index: sympy.Expr,
        dtype: torch.dtype,
        accu_store: bool = False,
    ):
        """
        Get a store line buffer that stores `value` into `var` at `index` of `dtype`. It handles
        both contiguous and non-contiguous store cases.
        :param value: Vectorized type templaterized on `dtype`.
        :param var: buffer to store into.
        :index: index into the `var`.
        """
        # when value's type is str (e.g., welford reduction), caller should make sure
        # it is a vector
        assert isinstance(value, str) or (
            isinstance(value, CppCSEVariable) and value.is_vec
        ), value
        tiling_var = self.itervars[self.tiling_idx]
        var_expr = f"{var} + {cexpr_index(index)}"
        stride = self._try_get_const_stride(index, tiling_var)
        code = IndentedBuffer()
        if stride == 1:
            if dtype == torch.float and self.tail_size is None:
                code.writeline(f"{value}.store({var_expr});")
            else:
                code.writeline(f"{value}.store({var_expr}, {self.num_elems});")
        else:
            self._load_or_store_non_contiguous(
                var, index, dtype, buffer=code, store_value=value, accu_store=accu_store
            )
        return code
```

Comment (jgong5) [reply torch/_inductor/codegen/cpp.py:2284]: If so, please add an assertion here to make sure the returned index is a vector: `assert index.is_vec`

**Review by jgong5 — APPROVED** [review]
Please add assertions as suggested. Others LGTM. We need to land it after the dependent PRs are landed.

**Review by leslie-fang-intel — APPROVED** [review]

**Pull request closed without merging**
